### DISCUSSION TITLE
pom file errors after SWT workspace is setup in Ubuntu using oomph

### DISCUSSION THREAD

**Discussion by lathapatil** [eclipse-platform/eclipse.platform.swt#1065] [Q&A] [answered]
I have set up an SWT workspace on the latest version of Ubuntu Linux and am able to run the Eclipse application locally after following the steps mentioned in  https://wiki.eclipse.org/SWT/Devel/Gtk/Dev_guide.

Now I'm attempting to modify the OS.java file and compile the new C code, following the steps outlined [here](https://www.eclipse.org/swt/jnigen.php)  but the Ant build scripts are no longer present in the workspace, and I'm attempting to trace the chain [here ](https://github.com/eclipse-platform/eclipse.platform.swt/issues/513#issuecomment-1899423505)

But the POM files in the workspace are having issues, and I'm wondering if it's a workspace setup problem. I haven't made any changes to the POM file.

![image](https://github.com/eclipse-platform/eclipse.platform.swt/assets/130035155/b6cda4a6-b7c3-48b3-bcad-0406dc162f73)

**Comment by merks** [comment]
I don't see that error...

I think the  direct parent is fine, but the  `<relativePath>../eclipse-platform-parent</relativePath>` it doesn't like

![image](https://github.com/eclipse-platform/eclipse.platform.swt/assets/208716/ac4c78f9-0989-4193-8c9d-8a71fb021c36)

I believe this error remains only as long as the editor is open...

The problem is that the overall pom structure assumes the repo is checked out as a submodule of 

https://github.com/eclipse-platform/eclipse.platform.releng.aggregator

So it expects that `eclipse-platform-parent` will be a sibling of the SWT repository root location:

![image](https://github.com/eclipse-platform/eclipse.platform.swt/assets/208716/fd37b5f8-ae27-4b01-817b-d5aef8e9dd8c)

**Comment by lathapatil** [comment]
Closing the POM editor resolved the issue today, but that wasn't the case when I was attempting this a couple of days ago.
Thanks for the tip.

Now I am proceeding with my Maven command to compile the new C code and copy the new libraries to the appropriate fragment

`/swt-master/git/eclipse.platform.swt/binaries$ mvn clean process-resources -Dnative=gtk.linux.x86_64 -DskipTests`

But I am encountering errors. Could it be a proxy issue? I am unable to locate the mentioned .pom file at the specified location.

 Original error: Could not transfer artifact org.eclipse.tycho:tycho-source-plugin:pom:4.0.6 from/to tycho-snapshots (https://repo.eclipse.org/content/repositories/tycho-snapshots/): HTTP proxy server authentication failed for https://repo.eclipse.org/content/repositories/tycho-snapshots/org/eclipse/tycho/tycho-source-plugin/4.0.6/tycho-source-plugin-4.0.6.pom 407 Proxy Authentication Required

When clicking on the .pom file link, I receive the following information:
Retrieval of /org/eclipse/tycho/tycho-source-plugin/4.0.6/tycho-source-plugin-4.0.6.pom from M2Repository(id=tycho-snapshots) is forbidden by repository policy SNAPSHOT.

**Reply by merks** [reply]
Tycho 4.0.6 is released so it's not present here:

https://repo.eclipse.org/content/repositories/tycho-snapshots/org/eclipse/tycho/tycho-source-plugin/

**Comment by laeubi** [comment]
@lathapatil If you are using a special maven setup you maybe need to configure the `settings.xml` or maybe use a different one (`-s` option) and configure them on CLI and in the IDE.

**Reply by lathapatil** [reply]
@merks  and @laeubi 

Thank you for your support.

_Am I taking the right approach by using Maven to compile *.c files?_
My understanding is that whenever we modify the OS.java file, SWT tools update the corresponding OS.c file, and then we need to execute a Maven command to build the new C code (since Ant scripts are no longer available) and to copy the new libraries to the appropriate fragment.

As I am new to SWT, I am having trouble understanding what's going wrong here, especially since I cannot find any updated documentation on compiling *.c files.

I have a settings.xml file where I've specified my proxy details exclusively. It might not be a proxy issue because when I execute the mentioned Maven command, I can successfully download some items..

![image](https://github.com/eclipse-platform/eclipse.platform.swt/assets/130035155/fa227b14-2719-4482-a291-f2546ac1b364)

I'm encountering an error related to the tycho-source-plugin-4.0.6.pom, and I noticed that the .mvn/extensions.xml file is pointing to version 4.0.4 for org.eclipse.tycho.

_Do I need to make any further changes in my local workspace to resolve this issue?_

![image](https://github.com/eclipse-platform/eclipse.platform.swt/assets/130035155/eb85ec16-f7e0-43cb-a955-256ca99ef997)

**Reply by laeubi** [reply]
Basically you should take a look here https://github.com/eclipse-platform/eclipse.platform.swt/blob/master/.github/workflows/maven.yml this is the "build as code" description of what we need in the PR, it basically boils down to install some prerequisites, make java versions available and then execute 

```
mvn --batch-mode -V -U
        -DforkCount=1
        -Dnative=gtk.linux.x86_64
        -Dcompare-version-with-baselines.skip=true
        -Dtycho.baseline.replace=none
        -DskipNativeTests=false
        -DfailIfNoTests=false
        clean verify
```

that should work if executed from the root of the git source tree (let your IDE closed just to make sure these do not interfere), especially `-U` is useful if a previous download attempt failed.

**Reply by akurtakov** [reply]
There is also https://github.com/eclipse-platform/eclipse.platform.swt/blob/master/bundles/org.eclipse.swt/Build-SWT-native-binaries-for-running-platform.launch which is supposed to automate that from the workbench.

**Reply by lathapatil** [reply]
From `eclipse.platform.swt/bundles/org.eclipse.swt/Readme.Linux.md`, I understand that I can either run the `Build-SWT-native-binaries-for-running-platform` Maven launch configuration or the `build.sh` script in the `binaries/org.eclipse.swt.gtk.linux.${arch}/bin/library` directory.

Will running `build.sh` also compile changed C files and copy the new libraries (*.so files) to the appropriate fragment?

If that is the case, do I no longer need to use Maven?

**Reply by HeikoKlare** [reply]
In case you specify the argument `install` when executing `build.sh`, it will also copy the generated *.so files:
https://github.com/eclipse-platform/eclipse.platform.swt/blob/d98c128328a826ffd48a03f576f2fdc007397d2f/bundles/org.eclipse.swt/Eclipse%20SWT%20PI/gtk/library/build.sh#L32

The Maven build also just executes the `build.sh` with according arguments. In addition, it defines some environment variables but if unspecified, most (if not all) of them are also filled with reasonable values by the `build.sh`.
https://github.com/eclipse-platform/eclipse.platform.swt/blob/d98c128328a826ffd48a03f576f2fdc007397d2f/binaries/pom.xml#L143-L150

**Reply by lathapatil** [reply]
@HeikoKlare  Thanks for the confirmation .
I can execute `build.sh` its always successful except that `OUTPUT_DIR `path should be updated manually from `OUTPUT_DIR=../../../../../eclipse.platform.swt/bundles/org.eclipse.swt.gtk.${SWT_OS}.${SWT_ARCH}` to `OUTPUT_DIR=../../../../../eclipse.platform.swt/binaries/org.eclipse.swt.gtk.${SWT_OS}.${SWT_ARCH}`

However, I still haven't been able to resolve or figure out how to do it using Maven. If you've already attempted to build it using Maven, could you please share the steps and commands with me?

**Reply by HeikoKlare** [reply, marked as answer]
I build the binaries via Maven from command line. A prerequisite for building the binaries is to have an `swtBuildQualifier.txt` in `bundles\org.eclipse.swt\target`, which is created by simply running an ordinary Maven build on the complete SWT repository (or you can probably also create it manually). Then you can execute the following within the folder of the fragment you want to build:
```
mvn clean process-resources -Dnative=${ws}.${os}.${arch} -DskipTests
```
So in your case it should be `mvn clean process-resources -Dnative=gtk.linux.x86_64 -DskipTests` in the `org.eclipse.swt.gtk.linux.x86_64` fragment folder. 

You can also find the description how to build the binaries here: https://github.com/eclipse-platform/eclipse.platform.swt/pull/973#issuecomment-1895398530

> I can execute `build.sh` its always successful except that `OUTPUT_DIR `path should be updated manually from `OUTPUT_DIR=../../../../../eclipse.platform.swt/bundles/org.eclipse.swt.gtk.${SWT_OS}.${SWT_ARCH}` to `OUTPUT_DIR=../../../../../eclipse.platform.swt/binaries/org.eclipse.swt.gtk.${SWT_OS}.${SWT_ARCH}`

True, the default output path is outdated. I was not probably not detected yet since the Maven build overwites it. Do you want to provide a PR to correct the path?

**Reply by lathapatil** [reply]
This really helped to resolve my issue with building SWT native binaries using the Maven approach. Thanks for the detailed steps.

>A prerequisite for building the binaries is to have an swtBuildQualifier.txt in bundles\org.eclipse.swt\target, which is created by simply running an ordinary Maven build on the complete SWT repository (or you can probably also create it manually).

If this step is mandatory, then it also has to be documented somewhere.

>True, the default output path is outdated. I was not probably not detected yet since the Maven build overwites it. Do you want to provide a PR to correct the path?

Sure , I can do that.
